frame/support/src/migrations.rs

```diff
@@ -15,19 +15,128 @@
+/// pub type Migrations = (
+/// 	// ...other migrations
+/// 	VersionedRuntimeUpgrade<
+/// 		Const16<4>, // From on-chain version 4
+/// 		Const16<5>, // To on-chain version 5
+/// 		parachains_configuration::migration::v5::MigrateToV5<Runtime>,
+/// 		Configuration,
+/// 		RocksDbWeight,
+/// 	>,
+/// 	VersionedRuntimeUpgrade<
+/// 		Const16<5>, // From on-chain version 5
+/// 		Const16<7>, // To on-chain version 7
+/// 		parachains_configuration::migration::v6::MigrateToV6<Runtime>,
+/// 		Configuration,
+/// 		RocksDbWeight,
+/// 	>,
+/// 	// ...other migrations
+/// );
+/// ```
+pub struct VersionedRuntimeUpgrade<From, To, Inner, Pallet, Weight> {
+	_marker: PhantomData<(From, To, Inner, Pallet, Weight)>,
+}
+
+/// Implementation of the `OnRuntimeUpgrade` trait for `VersionedRuntimeUpgrade`.
+///
+/// Its main function is to perform the runtime upgrade in `on_runtime_upgrade` only if the on-chain
+/// version of the pallets storage matches `From`, and after the upgrade set the on-chian storage to
+/// `To`. If the versions do not match, it writes a log notifying the developer that the migration
+/// is a noop.
+impl<
+		From: Get<u16>,
+		To: Get<u16>,
+		Inner: OnRuntimeUpgrade,
+		Pallet: GetStorageVersion<CurrentStorageVersion = StorageVersion> + PalletInfoAccess,
+		DbWeight: Get<RuntimeDbWeight>,
+	> OnRuntimeUpgrade for VersionedRuntimeUpgrade<From, To, Inner, Pallet, DbWeight>
+{
+	/// Executes post_upgrade if the migration will run.
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, sp_runtime::TryRuntimeError> {
+		let on_chain_version = Pallet::on_chain_storage_version();
+		if on_chain_version == From::get() {
+			Inner::pre_upgrade()
+		} else {
+			Ok(Vec::new())
+		}
+	}
+
+	/// Executes the versioned runtime upgrade.
+	///
+	/// First checks if the pallets on-chain storage version matches the version of this upgrade. If
+	/// it matches, it calls `Inner::on_runtime_upgrade`, updates the on-chain version, and returns
+	/// the weight. If it does not match, it writes a log notifying the developer that the migration
+	/// is a noop.
+	fn on_runtime_upgrade() -> Weight {
+		let on_chain_version = Pallet::on_chain_storage_version();
+		if on_chain_version == From::get() {
+			// Execute the migration
+			let weight = Inner::on_runtime_upgrade();
+
+			// Update the on-chain version
+			StorageVersion::new(To::get()).put::<Pallet>();
+
+			weight.saturating_add(DbWeight::get().reads_writes(1, 1))
+		} else {
+			log::warn!(
+				"{} VersionedOnRuntimeUpgrade {:?} skipped because current on-chain version is {:?}.",
+				Pallet::name(),
+				From::get(),
+				on_chain_version
+			);
+			DbWeight::get().reads(1)
+		}
+	}
+
+	/// Executes post_upgrade if the migration will run.
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(pre_upgrade_return_bytes: Vec<u8>) -> Result<(), sp_runtime::TryRuntimeError> {
+		let on_chain_version = Pallet::on_chain_storage_version();
+		if on_chain_version == From::get() {
```

Comment:
Is this correct? Should the on-chain version not be updated afterwards?  
Maybe you can just encode something like `(bool, inner)` and set the bool to whether `pre_upgrade` executed.

Reply:
Good catch - thanks.